### PR TITLE
Fix exception causes in most of the codebase

### PR DESCRIPTION
**Pull request by cool-RR**
@rgommers This is a continuation of #12353

### CODE REVIEW

**Comment by cool-RR** [comment]
I assume that the test failure is not related to my change, let me know otherwise.

File: scipy/interpolate/interpolate.py
Comment (rlucas7) [inline scipy/interpolate/interpolate.py:224]: since you're looking at these changes anyway could we make the changes be in favor of improving the information content of the error messages? 

for example in this one, rather than write: 
`ValueError("Unsupported interpolation type.")`
it could read 
`ValueError("{0} Unsupported interpolation type {1}.".format(key, type(key))` ?

The benefit is that someone reading error message (or logs) will determine that `key` has type `X` (whatever) and that is not supported for interpolation. Whereas as this PR reads, user only learns that there is a type w/unsupported interpolation (but does not know what that unsupported type is). 

Thoughts?

File: scipy/interpolate/interpolate.py
Comment (rlucas7) [inline scipy/interpolate/interpolate.py:1717]: similar comment here, I feel like a fair bit of the error messaging in SciPy is written assuming an interactive session is being used. This is definitely an important use case but certainly not the only way SciPy is used. 

To be clear, I am not suggesting ***all*** of the error messages in this PR should be changes to include interpolation but some would clearly aid the end user in root causing source of the raised error.

**Review by rlucas7 — CHANGES_REQUESTED** [review]
@cool-RR I took a look, as (last I knew) @rgommers is on holiday for a few (weeks). 

The changes here all look good and will cleanup the emitted error messages, thanks for these. 

I left one suggestion if you're up for it, would require maybe 1 more pass through the messages, if you agree.

File: scipy/interpolate/interpolate.py
Comment (cool-RR) [reply scipy/interpolate/interpolate.py:224]: That sounds like a good direction, but since this is a change in many files across the repo, I think we should be extra conservative. Any improvements should be made on separate PRs that are either module-specific or specific to the type of change being done.

File: scipy/interpolate/interpolate.py
Comment (cool-RR) [reply scipy/interpolate/interpolate.py:1717]: Same answer as the other comment.

File: scipy/interpolate/interpolate.py
Comment (rlucas7) [reply scipy/interpolate/interpolate.py:224]: > That sounds like a good direction, but since this is a change in many files across the repo, I think we should be extra conservative. Any improvements should be made on separate PRs that are either module-specific or specific to the type of change being done.

Fair enough, perhaps you are interested to open (some of) those PR(s)?

**Review by rlucas7 — APPROVED** [review]
These changes all seem pretty straightforward so marking as approved. 
Given that the changes in the PR are large I'll mark approved by leave open for the week in case anyone wants to comment or request further changes. If none, I'll merge next weekend. 

EDIT: by large changes, I mean the changes touch many sub-packages, the changes themselves are all very straightforward changes to the error handling to have cleaner messaging.

File: scipy/interpolate/interpolate.py
Comment (cool-RR) [reply scipy/interpolate/interpolate.py:224]: Yes. Would you be okay with f-strings instead of `.format`? Since SciPy requires Python 3.6 I think these are okay now. Also, I assume you meant after this PR is merged, because otherwise we'll have a merge conflict.

Comment (rlucas7) [reply scipy/interpolate/interpolate.py:224]: > Yes. Would you be okay with f-strings instead of `.format`? Since SciPy requires Python 3.6 I think these are okay now. 

I'm not sure what the preference is on f-string versus, `.format` the scipy code base has both "old style" interpolation w/ `%` as well as the `.format`. I'm not aware of any use of f-strings but my github searching skills are not at the level where a query on `f"` or `f\"` will return anything for the repo. I'll ask on the dev-list what the preference is for the string interpolation and post here once reply. 

>Also, I assume you meant after this PR is merged, because otherwise we'll have a merge conflict.

Either way is fine, merge conflicts can be resolved on a rebase or you could wait until this one is merged.

Comment (cool-RR) [reply scipy/interpolate/interpolate.py:224]: I understand. I'll wait until we're done with fixing the exception causes, then open the separate PR.

Comment (rgommers) [reply scipy/interpolate/interpolate.py:224]: Agree with this conversation: separate PR for follow-ups, and using f-strings is fine for new code or code you're touching anyway.

**Review by rgommers — APPROVED** [review]
LGTM too, thanks @cool-RR. And thanks @rlucas7 for the detailed review.

**Pull request merged**
